src/globus_sdk/experimental/auth_requirements_error/_variants.py

```diff
@@ -0,0 +1,309 @@
+from __future__ import annotations
+
+import typing as t
+
+from . import validators
+from .auth_requirements_error import (
+    GlobusAuthorizationParameters,
+    GlobusAuthRequirementsError,
+)
+
+T = t.TypeVar("T", bound="LegacyAuthRequirementsErrorVariant")
+
+
+class LegacyAuthRequirementsErrorVariant:
+    """
+    Abstract base class for errors which can be converted to a
+    Globus Auth Requirements Error.
+    """
+
+    SUPPORTED_FIELDS: dict[str, t.Callable[[t.Any], t.Any]] = {}
+
+    @classmethod
+    def from_dict(cls: t.Type[T], error_dict: dict[str, t.Any]) -> T:
+        """
+        Instantiate from an error dictionary.
+
+        :param error_dict: The dictionary to instantiate the error from.
+        :type error_dict: dict
+        """
+        # Extract any extra fields
+        extras = {k: v for k, v in error_dict.items() if k not in cls.SUPPORTED_FIELDS}
+        kwargs: dict[str, t.Any] = {"extra": extras}
+        # Ensure required fields are supplied
+        for field_name in cls.SUPPORTED_FIELDS.keys():
+            kwargs[field_name] = error_dict.get(field_name)
+
+        return cls(**kwargs)
+
+    def to_auth_requirements_error(self) -> GlobusAuthRequirementsError:
+        raise NotImplementedError()
```

Comment:
I would like `LegacyAuthRequirementsErrorVariant` to be a protocol class which the other classes implement, but without providing any concrete methods.
Because `from_dict` is used in the `utils.py` module, it needs to be declared on the protocol, but the following is a suitable definition:

```python
class LegacyAuthRequirementsErrorVariant(Protocol):
    @classmethod
    def from_dict(cls: t.Type[T], error_dict: dict[str, t.Any]) -> T:
        ...

    def to_auth_requirements_error(self) -> GlobusAuthRequirementsError:
        ...
```

The main reason to prefer this declaration is that no inheritance is used, but type checkers will require that the type is matched in any assignment. For example, the following assignment requires (under type checkers) that the protocol is satisfied by the implementing class:
```python
x: type[LegacyAuthRequirementsErrorVariant] = LegacyConsentRequiredTransferError
```

This also separates out a question which I think should be answered differently:

I think we should make the implementations of `from_dict` and `to_dict` common for all of our classes. (I recognize that the classes in this module don't define `to_dict`, but I would have them inherit it.)
Rather than try to describe in detail, I'll just point to what I've done with `Serializable` in #796. Other names are fine, but I think sharing a common base for all of these helper objects which is defined _as_ a common base and is non-abstract leads to a better model.